tree/ntuple/v7/inc/ROOT/RNTupleWriter.hxx

```diff
@@ -0,0 +1,271 @@
+/// \file ROOT/RNTupleWriter.hxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2024-02-20
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2024, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleWriter
+#define ROOT7_RNTupleWriter
+
+#include <ROOT/RConfig.hxx> // for R__unlikely
+#include <ROOT/REntry.hxx>
+#include <ROOT/RError.hxx>
+#include <ROOT/RNTupleMetrics.hxx>
+#include <ROOT/RNTupleModel.hxx>
+#include <ROOT/RNTupleUtil.hxx>
+#include <ROOT/RPageStorage.hxx>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string_view>
+#include <utility>
+
+class TFile;
+
+namespace ROOT {
+namespace Experimental {
+
+class RNTupleWriteOptions;
+
+// clang-format off
+/**
+\class ROOT::Experimental::RNTupleFillContext
+\ingroup NTuple
+\brief A context for filling entries (data) into clusters of an RNTuple
+
+An output cluster can be filled with entries. The caller has to make sure that the data that gets filled into a cluster
+is not modified for the time of the Fill() call. The fill call serializes the C++ object into the column format and
+writes data into the corresponding column page buffers.  Writing of the buffers to storage is deferred and can be
+triggered by CommitCluster() or by destructing the context.  On I/O errors, an exception is thrown.
+
+Instances of this class are not meant to be used in isolation and can be created from an RNTupleParallelWriter. For
+sequential writing, please refer to RNTupleWriter.
+*/
+// clang-format on
+class RNTupleFillContext {
```

Comment:
Maybe we want `RNTupleFillContext` in its own header as well? (The parallel writer uses it as well and it's a bit weird to `#include <ROOT/RNTupleWriter.hxx>` for it...)

Reply:
Yep, I'll do this in a follow-up PR.